rm: Implement --one-file-system and --preserve-root=all

shskwmt opened the pull request:
Fixes https://github.com/uutils/coreutils/issues/7011

Implement `--one-file-system` and `--preserve-root=all` options for the `rm` command.

sylvestre commented on src/uu/rm/src/rm.rs:
```diff
@@ -482,9 +502,91 @@ fn remove_dir_recursive(path: &Path, options: &Options) -> bool {
+        return true;
+    }
+
+    match validate_single_filesystem(path) {
+        Ok(()) => true,
+        Err(additional_reason) => {
+            if !additional_reason.is_empty() {
+                show_error!("{}", additional_reason);
+            }
+            show_error!(
+                "skipping {}, since it's on a different device",
+                path.quote()
+            );
+            if options.preserve_root == PreserveRoot::YesAll {
+                show_error!("and --preserve-root=all is in effect");
+            }
+            false
+        }
+    }
```
can probably be simplified with something like:
```suggestion
let result = validate_single_filesystem(path);
if result.is_ok() {
    return true;
}

if let Err(additional_reason) = result {
    if !additional_reason.is_empty() {
        show_error!("{}", additional_reason);
    }
}

show_error!(
    "skipping {}, since it's on a different device",
    path.quote()
);

if options.preserve_root == PreserveRoot::YesAll {
    show_error!("and --preserve-root=all is in effect");
}

false
```

shskwmt replied:
Thank you for pointing that out! I've implemented the suggested simplification.

https://github.com/uutils/coreutils/pull/7569/commits/c0913696b976677ec2d0e8f62a92d1271acac554

sylvestre commented:
some jobs are failing :)

sylvestre commented:
does
 bash -v util/run-gnu-test.sh tests/rm/one-file-system.sh  
passes on your system ?

shskwmt commented:
The following is the test result. It has passed.

```
PASS: tests/rm/one-file-system.sh
============================================================================
Testsuite summary for GNU coreutils 9.6-dirty
============================================================================
# TOTAL: 1
# PASS:  1
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
```

shskwmt commented:
I tried again and it failed, so I will fix it.

```
FAIL: tests/rm/one-file-system
==============================

--- exp 2025-03-31 00:11:32.336030079 +0900
+++ out 2025-03-31 00:11:32.336030079 +0900
@@ -1 +1 @@
-rm: skipping 'a', since it's on a different device
+rm: cannot remove 'a/b': Device or resource busy
FAIL tests/rm/one-file-system.sh (exit status: 1)

============================================================================
Testsuite summary for GNU coreutils 9.6-dirty
============================================================================
# TOTAL: 1
# PASS:  0
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0
============================================================================
See ./tests/test-suite.log for debugging.
Some test(s) failed.  Please report this to bug-coreutils@gnu.org,
together with the test-suite.log file (gzipped) and your system
information.  Thanks.
============================================================================
```

shskwmt commented:
No warries, I'll fix it.
Thanks for reaching out!

sylvestre commented:
sorry, i missed it :( 
it needs to be rebased, could you please do this? sorry :(

shskwmt committed:
feat(rm): Implement --one-file-system and --preserved-root=all

This comment enhances the safety of recursive deletion in `rm` by introducing two key features to
prevent accidental data loss.

`--one-file-system`: When specified, `rm` will not traverse into directories that are on a
different file system from the one on which the traversal began, this prevents `rm -r` from
accidentally deleting data on mounted volumes.

`--preserve-root=all`: The `--preserve-root` option is enhanced to accept an optional argument,
`all`. When set to `all`, `rm` will refuse to remove any directory that is a mount point. The
default behavior (`--preserve-root` without an argument) remains, protecting only the root
directory (`/`).

sylvestre commented on src/uu/rm/src/rm.rs:
```diff
@@ -242,7 +250,18 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
+        } else {
+            match matches
+                .get_one::<String>(OPT_PRESERVE_ROOT)
+                .unwrap()
```
please remove the unwrap()

shskwmt replied:
Fixed:
https://github.com/uutils/coreutils/pull/7569/commits/a0df78a4e0ba51c486214701b00a417020fc9589

sylvestre commented on src/uu/rm/src/rm.rs:
```diff
@@ -684,9 +717,82 @@ fn remove_dir_recursive(
+        .map_err(|err| format!("cannot canonicalize {}: {err}", path.quote()))?;
+
+    // Get parent path, handling root case
+    let parent_canon = child_canon.parent().ok_or("")?.to_path_buf();
```
can we avoid the empty strings in ok_or(") ?

shskwmt replied:
Fixed:
https://github.com/uutils/coreutils/pull/7569/changes/2617af7968cf11089244d5087ab2e61351a1fe13

sylvestre commented on src/uu/rm/src/rm.rs:
```diff
@@ -684,9 +717,82 @@ fn remove_dir_recursive(
 fn handle_dir(path: &Path, options: &Options, progress_bar: Option<&ProgressBar>) -> bool {
     let mut had_err = false;
 
+    if let Err(additional_reason) = check_one_fs(path, options) {
```
same code as in line 621, please dedup

sylvestre committed:
Merge branch 'main' into rm-one-file-system

ChrisDryden commented:
I see the Smack tests are failing in the CI here and it appears that its due to no space left on device. I'm guessing its caused when downloading the arch linux img

sylvestre committed:
Merge branch 'main' into rm-one-file-system

sylvestre committed:
Merge branch 'main' into rm-one-file-system

shskwmt committed:
refactor(rm): improve preserve-root parsing safety

- Replace `.unwrap()` with `.map()` to handle options safely without risking panics.
- Update the abbreviation check to compare against the `PreserveRoot` enum directly.

shskwmt committed:
refactor(rm): improve error context in check_one_fs

Replace empty error strings with descriptive messages when path parent
or mount point resolution fails. This aids in debugging by identifying
specifically which path caused the resolution failure during the
one-file-system check.

shskwmt committed:
refactor(rm): extract one-fs and root preservation checks to helper

Introduce `check_and_report_one_fs` to encapsulate the logic for
verifying filesystem boundaries (`-x`) and root preservation.

This removes code duplication between `remove_dir_recursive` and
`handle_dir`, ensuring consistent error reporting for "different device"
skips and `--preserve-root=all` violations.

shskwmt committed:
refactor(df): separate platform-specific logic for stat_path

Split the initialization of `stat_path` in `Filesystem::new` into distinct
`#[cfg(unix)]` and `#[cfg(windows)]` blocks. This improves readability and
ensures that Windows builds consistently use the volume ID (`dev_id`).

sylvestre committed:
Merge branch 'main' into rm-one-file-system

ChrisDryden commented:
It appears that one of the error messages is returning differently in the test:

```
--- exp	2026-01-21 10:59:43.543481609 +0000
+++ out	2026-01-21 10:59:43.546481572 +0000
@@ -1 +1 @@
-rm: skipping 'a/b', since it's on a different device
+rm: cannot remove 'a/b': Device or resource busy
FAIL tests/rm/one-file-system.sh (exit status: 1)
```

shskwmt commented:
@ChrisDryden 
Hi, sorry for the late reply.

I actually have a fix ready locally, but I haven't been able to push it yet as I'm still in the middle of testing. It's a bit tough for me to find time during the week, so I'll likely get to it over the weekend.

For reference, here is the current diff I'm working on.

If this is urgent, please feel free to cherry-pick and modify it as you suggested!

```
diff --git a/src/uu/rm/src/platform/unix.rs b/src/uu/rm/src/platform/unix.rs
index e890ab158..a0092e742 100644
--- a/src/uu/rm/src/platform/unix.rs
+++ b/src/uu/rm/src/platform/unix.rs
@@ -11,6 +11,7 @@ use indicatif::ProgressBar;
 use std::ffi::OsStr;
 use std::fs;
 use std::io::{IsTerminal, stdin};
+use std::os::unix::fs::MetadataExt;
 use std::os::unix::fs::PermissionsExt;
 use std::path::Path;
 use uucore::display::Quotable;
@@ -21,9 +22,9 @@ use uucore::show_error;
 use uucore::translate;
 
 use super::super::{
-    InteractiveMode, Options, is_dir_empty, is_readable_metadata, prompt_descend, remove_file,
-    show_permission_denied_error, show_removal_error, verbose_removed_directory,
-    verbose_removed_file,
+    InteractiveMode, Options, PreserveRoot, is_dir_empty, is_readable_metadata, prompt_descend,
+    remove_file, show_one_fs_error, show_permission_denied_error, show_removal_error,
+    verbose_removed_directory, verbose_removed_file,
 };
 
 #[inline]
@@ -279,11 +280,11 @@ pub fn safe_remove_dir_recursive(
 ) -> bool {
     // Base case 1: this is a file or a symbolic link.
     // Use lstat to avoid race condition between check and use
-    let initial_mode = match fs::symlink_metadata(path) {
+    let (initial_mode, initial_dev) = match fs::symlink_metadata(path) {
         Ok(metadata) if !metadata.is_dir() => {
             return remove_file(path, options, progress_bar);
         }
-        Ok(metadata) => metadata.permissions().mode(),
+        Ok(metadata) => (metadata.permissions().mode(), metadata.dev()),
         Err(e) => {
             return show_removal_error(e, path);
         }
@@ -309,7 +310,7 @@ pub fn safe_remove_dir_recursive(
         }
     };
 
-    let error = safe_remove_dir_recursive_impl(path, &dir_fd, options);
+    let error = safe_remove_dir_recursive_impl(path, &dir_fd, options, Some(initial_dev));
 
     // After processing all children, remove the directory itself
     if error {
@@ -346,7 +347,12 @@ pub fn safe_remove_dir_recursive(
 }
 
 #[cfg(not(target_os = "redox"))]
-pub fn safe_remove_dir_recursive_impl(path: &Path, dir_fd: &DirFd, options: &Options) -> bool {
+pub fn safe_remove_dir_recursive_impl(
+    path: &Path,
+    dir_fd: &DirFd,
+    options: &Options,
+    parent_dev: Option<u64>,
+) -> bool {
     // Read directory entries using safe traversal
     let entries = match dir_fd.read_dir() {
         Ok(entries) => entries,
@@ -380,6 +386,15 @@ pub fn safe_remove_dir_recursive_impl(path: &Path, dir_fd: &DirFd, options: &Opt
         let is_dir = ((entry_stat.st_mode as libc::mode_t) & libc::S_IFMT) == libc::S_IFDIR;
 
         if is_dir {
+            if options.one_fs || options.preserve_root == PreserveRoot::YesAll {
+                if let Some(p_dev) = parent_dev {
+                    if entry_stat.st_dev as u64 != p_dev {
+                        show_one_fs_error(&entry_path, options);
+                        continue;
+                    }
+                }
+            }
+
             // Ask user if they want to descend into this directory
             if options.interactive == InteractiveMode::Always
                 && !is_dir_empty(&entry_path)
@@ -408,8 +423,12 @@ pub fn safe_remove_dir_recursive_impl(path: &Path, dir_fd: &DirFd, options: &Opt
                 }
             };
 
-            let child_error = safe_remove_dir_recursive_impl(&entry_path, &child_dir_fd, options);
-            error |= child_error;
+            let child_error = safe_remove_dir_recursive_impl(
+                &entry_path,
+                &child_dir_fd,
+                options,
+                Some(entry_stat.st_dev as u64)
+            );
 
             // Ask user permission if needed for this subdirectory
             if !child_error
@@ -435,7 +454,12 @@ pub fn safe_remove_dir_recursive_impl(path: &Path, dir_fd: &DirFd, options: &Opt
 }
 
 #[cfg(target_os = "redox")]
-pub fn safe_remove_dir_recursive_impl(_path: &Path, _dir_fd: &DirFd, _options: &Options) -> bool {
+pub fn safe_remove_dir_recursive_impl(
+    _path: &Path,
+    _dir_fd: &DirFd,
+    _options: &Options,
+    _parent_dev: Option<u64>,
+) -> bool {
     // safe_traversal stat_at is not supported on Redox
     // This shouldn't be called on Redox, but provide a stub for compilation
     true // Return error
diff --git a/src/uu/rm/src/rm.rs b/src/uu/rm/src/rm.rs
index 9f030ea1e..2bdd4805f 100644
--- a/src/uu/rm/src/rm.rs
+++ b/src/uu/rm/src/rm.rs
@@ -586,6 +586,17 @@ fn is_writable_metadata(_metadata: &Metadata) -> bool {
     true
 }
 
+pub(crate) fn show_one_fs_error(path: &Path, options: &Options) {
+    show_error!(
+        "skipping {}, since it's on a different device",
+        path.quote()
+    );
+
+    if options.preserve_root == PreserveRoot::YesAll {
+        show_error!("and --preserve-root=all is in effect");
+    }
+}
+
 /// Helper function to check fs and report errors if necessary.
 /// Returns true if the operation should be skipped/returned (i.e., on error).
 fn check_and_report_one_fs(path: &Path, options: &Options) -> bool {
@@ -593,14 +604,7 @@ fn check_and_report_one_fs(path: &Path, options: &Options) -> bool {
         if !additional_reason.is_empty() {
             show_error!("{}", additional_reason);
         }
-        show_error!(
-            "skipping {}, since it's on a different device",
-            path.quote()
-        );
-
-        if options.preserve_root == PreserveRoot::YesAll {
-            show_error!("and --preserve-root=all is in effect");
-        }
+        show_one_fs_error(path, options);
 
         return true;
     }
```

shskwmt committed:
feat(rm): implement one-file-system check for safe recursive removal

- Track device IDs (st_dev) during safe directory traversal to respect
  `--one-file-system` and `--preserve-root=all` options.
- Pass the parent device ID through recursive calls in `safe_remove_dir_recursive_impl`.
- Refactor `show_one_fs_error` into a reusable function to unify error reporting.
- Ensure the parent directory is not removed if any child removal fails or is skipped.

shskwmt committed:
fix(rm): fix error reporting when user declines interactive removal

Previously, `safe_remove_dir_recursive` used a single boolean to track
both system errors and skipped files. This meant that if a user
declined to remove a file interactively, it was treated internally as
an error.

This change splits the state into `cmd_error` and `child_remains`.
Now, declining a file prevents the parent directory from being removed
(because it is not empty) but does not propagate a failure exit code.

shskwmt committed:
fix(rm): set child_remains only when subdirectory removal actually fails

This refines the recursive removal logic on Unix to avoid incorrectly
preventing parent directory removal.

Previously, any failure to open a subdirectory (e.g., due to EACCES)
would unconditionally set `child_remains = true`. However, if the
directory is unreadable but empty, `handle_permission_denied` may
successfully unlink it.

Now, we only set `child_remains` and `cmd_error` if the direct removal
attempt (as a fallback for the open failure) actually fails. This fixes
the `test_inaccessible_dir_recursive` test case where a parent
directory remained even though its unreadable child was successfully
removed.

sylvestre commented on src/uu/rm/src/rm.rs:
```diff
@@ -564,6 +586,31 @@ fn is_writable_metadata(_metadata: &Metadata) -> bool {
+        path.quote()
+    );
+
+    if !options.one_fs && options.preserve_root == PreserveRoot::YesAll {
```
preserve-root message should show when preserve_root == PreserveRoot::YesAll, not when both conditions are false

no ?

sylvestre commented on src/uu/rm/src/rm.rs:
```diff
@@ -673,9 +725,75 @@ fn remove_dir_recursive(
+    }
+
+    // Read mount information
+    let fs_list = read_fs_list().map_err(|err| format!("cannot read mount info: {err}"))?;
```
Reading mount table for every file check introduces race conditions and poor performance

shskwmt committed:
refactor(rm): rename variables in unix platform for clarity

Rename `cmd_error` to `had_command_error` and `child_remains` to
`has_remaining_children` in `platform/unix.rs`.

These new names better reflect the boolean nature of the variables and
improve the readability of the control flow.

shskwmt committed:
refactor(rm): --one-file-system to use metadata-based device ID check

- Replace path canonicalization and mount point scanning with a more
  efficient stat-based (device ID) check.
- Add Windows support using volume_serial_number.
- Introduce OneFsError enum for clearer error reporting.
- Remove unused mount_for_path and read_fs_list.

shskwmt committed:
fix(rm): Windows build by using stable path prefix for device check

- Replace unstable volume_serial_number() with std::path::Component::Prefix.
  This allows the filesystem boundary check to work on stable Rust while
  correctly identifying drive crossings on Windows.
- Remove unused uucore::fsext::MountInfo import.

shskwmt committed:
rm: fix unused metadata and enum variant warnings on Windows

This patch fixes several warnings on Windows by:
- Moving Unix-specific metadata retrieval into the `#[cfg(unix)]` block.
- Guarding the `StatFailed` error variant and its handling with `#[cfg(unix)]`.
- Removing the unused `MetadataExt` import for Windows.

shskwmt committed:
Merge branch 'main' into rm-one-file-system
